fonts/generate.py

```diff
@@ -0,0 +1,657 @@
+Please do not use directly.
+"""
+
+log = logging.getLogger(__name__)
+
+
+def generate_smufl(opts: Namespace) -> bool:
+    """
+    Generates the `smufl.h` file for Verovio.
+
+    :param opts: A set of options from the argument parser sub-command.
+    :return: True if successful, False otherwise.
+    """
+    supported_glyphs_fname: str = opts.supported
+    header_out_dir: str = opts.header_out
+    supported_path: Path = Path(supported_glyphs_fname)
+    header_out_path: Path = Path(header_out_dir)
+    header_file_path: Path = Path(header_out_path, "smufl.h")
+
+    if not supported_path.is_file() or not os.access(supported_path, os.R_OK):
+        log.error(f"Could not find or read {supported_glyphs_fname} file.")
+        return False
+
+    if not os.access(header_out_path, os.W_OK):
+        log.error(f"Could not write to {header_out_dir}.")
+        return False
+
+    log.debug("SMuFL header will be stored in %s", header_out_path)
+    supported_glyphs: dict = __get_supported_glyph_codes(supported_glyphs_fname)
+
+    fmt_supported_glyphs: list = [
+        f"    SMUFL_{gcode}_{gname} = 0x{gcode}," for gcode, gname in supported_glyphs.items()
+    ]
+
+    log.debug("SMuFL header will contain %s supported glyphs", len(fmt_supported_glyphs))
+    fmt_glyph_list = "\n".join(fmt_supported_glyphs)
+    fmt_header = SMUFL_HEADER.format(
+        smufl_glyph_list=fmt_glyph_list, len_smufl_codes=len(fmt_supported_glyphs)
+    )
+
+    log.debug("Writing %s", header_file_path)
+    with open(header_file_path, "w") as header_inc:
+        header_inc.write(fmt_header)
+
+    log.debug("Finished writing SMuFL header")
+    return True
+
+
+def extract_fonts(opts: Namespace) -> bool:
+    """
+    Takes a font file and extracts the necessary files for Verovio. Generates a glyph file for each glyph, and an SVG
+    glyph file containing only the supported SMuFL glyphs.
+
+    :param opts: A set of options from the argument parser sub-command.
+    :return: True if successful, False otherwise.
+    """
+    source_dir: str = opts.source
+    fontname: str = opts.fontname
+    data_dir: str = opts.data
+
+    source_pth: Path = Path(source_dir)
+    font_data_pth: Path = Path(source_pth, fontname)
+    metadata_pth: Path = Path(font_data_pth, f"{fontname.lower()}_metadata.json")
+    font_pth: Path = Path(source_pth, font_data_pth, f"{fontname}.svg")
+    data_pth: Path = Path(data_dir)
+    glyph_file_pth: Path = Path(data_pth, fontname)
+    output_pth: Path = Path(data_pth, f"{fontname}.xml")
+
+    log.debug("Extracting fonts for %s from %s", fontname, font_data_pth.resolve())
+
+    if not font_data_pth.is_dir() or not os.access(font_data_pth, os.R_OK):
+        log.error(f"Could not read font information from {str(font_data_pth)}")
+        return False
+
+    if not os.access(metadata_pth, os.R_OK):
+        log.error(f"Could not read {str(metadata_pth)}. Does it exist?")
+        return False
+
+    if not os.access(font_pth, os.R_OK):
```

Comment:
Why having all these if statements for checking access. I think it would be more elegant to catch a `PermissionError` in `__read_svg_font_file` and log the error there.

Reply:
Because there are several different failure modes being checked. `PermissionError` only checks the permission, while os.R_OK checks both permission and existence.